p2p: track and use all potential peers for orphan resolution

glozow opened the pull request:
Part of #27463.

(Transaction) **orphan resolution** is a process that kicks off when we are missing UTXOs to validate an unconfirmed transaction. We currently request missing parents by txid; BIP 331 also defines a way to [explicitly request ancestors](https://github.com/bitcoin/bips/blob/master/bip-0331.mediawiki#handle-orphans-better).

Currently, when we find that a transaction is an orphan, we only try to resolve it with the peer who provided the `tx`. If this doesn't work out (e.g. they send a `notfound` or don't respond), we do not try again. We actually can't, because we've already forgotten who else could resolve this orphan (i.e. all the other peers who announced the transaction).

What is wrong with this? It makes transaction download less reliable, particularly for 1p1c packages which must go through orphan resolution in order to be downloaded.

Can we fix this with BIP 331 / is this "duct tape" before the real solution?
BIP 331 (receiver-initiated ancestor package relay) is also based on the idea that there is an orphan that needs resolution, but it's just a new way of communicating information. It's not inherently more honest; you can request ancestor package information and get a `notfound`. So ancestor package relay still requires some kind of procedure for retrying when an orphan resolution attempt fails. See the #27742 implementation which builds on this orphan resolution tracker to keep track of what packages to download (it just isn't rebased on this exact branch). The difference when using BIP 331 is that we request `ancpkginfo` and then `pkgtxns` instead of the parent txids.

Zooming out, we'd like orphan handling to be:
- Bandwidth-efficient: don't have too many requests out at once. As already implemented today, transaction requests for orphan parents and regular download both go through the `TxRequestTracker` so that we don't have duplicate requests out.
- Not vulnerable to censorship: don't give up too easily, use all candidate peers. See e.g. https://bitcoincore.org/en/2024/07/03/disclose_already_asked_for/
- Load-balance between peers: don't overload peers; use all peers available. This is also useful for when we introduce per-peer orphan protection, since each peer will have limited slots.

The approach taken in this PR is to think of each peer who announces an orphan as a potential "orphan resolution candidate." These candidates include:
- the peer who sent us the orphan tx
- any peers who announced the orphan prior to us downloading it
- any peers who subsequently announce the orphan after we have started trying to resolve it
For each orphan resolution candidate, we treat them as having "announced" all of the missing parents to us at the time of receipt of this orphan transaction (or at the time they announced the tx if they do so after we've already started tracking it as an orphan). We add the missing parents as entries to `m_txrequest`, incorporating the logic of typical txrequest processing, which means we prefer outbounds, try not to have duplicate requests in flight, don't overload peers, etc.

glozow committed:
[txrequest] GetCandidatePeers

Needed for a later commit adding logic to ask the TxRequestTracker for a
list of announcers.  These announcers should know the parents of the
transaction they announced.

glozow committed:
[txorphanage] support multiple announcers

Add ability to add and track multiple announcers per orphan transaction,
erasing announcers but not the entire orphan.

The tx creation code in orphanage_tests needs to be updated so that each
tx is unique, because the CountOrphans() check assumes that calling
EraseForPeer necessarily means its orphans are deleted.

Unused for now.

glozow committed:
[refactor] change type of unique_parents to Txid

glozow committed:
[p2p] try multiple peers for orphan resolution

Co-authored-by: dergoegge <n.goeggi@gmail.com>

glozow committed:
[functional test] orphan handling with multiple announcers

glozow committed:
[unit test] TxOrphanage EraseForBlock

glozow committed:
[fuzz] orphanage multiple announcer functions

glozow committed:
[p2p] only attempt 1p1c when both txns provided by the same peer

Now that we track all announcers of an orphan, it's not helpful to
consider an orphan provided by a peer that didn't send us this parent.
It can only hurt our chances of finding the right orphan when there are
multiple candidates.

Adapt the 2 tests in p2p_opportunistic_1p1c.py that looked at 1p1c
packages from different peers. Instead of checking that the right peer
is punished, we now check that the package is not submitted. We can't
use the functional test to see that the package was not considered
because the behavior is indistinguishable (except for the logs).

glozow committed:
[txdownload] remove unique_parents that we already have

This means we no longer return parents we already have in the
m_unique_parents result from MempoolRejectedTx.

We need to separate the loop that checks AlreadyHave parents from the
loop that adds parents as announcements, because we may do the latter
loop multiple times for different peers.

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -204,8 +227,83 @@ bool TxDownloadManagerImpl::AddTxAnnouncement(NodeId peer, const GenTxid& gtxid,
+        // existing behavior: drop if we are tracking too many invs for this peer already. Each
+        // orphan resolution involves at least 1 transaction request which may or may not be
+        // currently tracked in m_txrequest, so we include that in the count.
+        if (m_txrequest.Count(nodeid) + m_orphan_resolution_tracker.Count(nodeid) >= MAX_PEER_TX_ANNOUNCEMENTS) return std::nullopt;
```
Ah, need to add the size of `unique_parents` to this. Surprised fuzzer has not tripped on it...

dergoegge reviewed (commented):
Concept ACK

I implemented what I was talking about in the [review club yesterday](https://bitcoincore.reviews/31397) here: https://github.com/dergoegge/bitcoin/commits/2024-12-rslv-orphns/ (the only difference is the main commit 3928b400d805f75e0ba5b9947d5d441a828aa5be) and it looks like it passes your functional test additions! So perhaps it's worth exploring if we actually need another tracker...


<details>

<summary>IRC logs</summary>

```
17:34 	<dergoegge> I was wondering why we need another txrequest tracker instance for this? couldn’t we just keep track of announcers in the orphanage (same as in the PR) and then add requests for the parents to the existing tracker on future announcements?
17:35 	<glozow> dergoegge: that's an idea. How would you schedule the requests?
17:35 	<dergoegge> ideally at the same time
17:35 	<glozow> Is there a cost to having another txrequest tracker? It's not that different from adding another std::map
17:35 	<glozow> No, I mean, how would you load-balance between peers, bake in preferences, etc.
17:36 	<dergoegge> isn't that what the existing tracker does?
17:36 	<glozow> Oh, you mean adding a new type to the tracker? So we'd have txid type, wtxid type, and orphan?
17:37 	<glozow> also note that the parent requests are added to the txrequest tracker
17:39 	<dergoegge> I guess I'm wondering why we need the concept of tracking the resolution by orphan id, as opposed to just putting the requests for the parents in the existing tracker
17:39 	<glozow> we do put the requests for the parents in the existing tracker
17:39 	<glozow> Maybe we are crossing wires?
17:39 	<instagibbs> mmmm he's asking why the add to new tracker, then "immediately" add to old one, vs just add to old one, I think
17:40 	<dergoegge> yea
17:40 	<instagibbs> add to old one with "proper delays"
17:40 	<instagibbs> I didn't get far enough in my review to figure this out either 😅
17:40 	<glozow> We might have multiple candidates for orphan resolution
17:41 	<glozow> Oh I see what you're saying
17:42 	<dergoegge> "multiple candidates" as in same parents different orphan?
17:42 	<glozow> Perhaps that could work, where you're treating it as if all of them just announced the missing parents? I don't know how you'd add `ancpkginfo` orphan resolution easily this way though.
17:43 	<glozow> different peers same orphan
17:43 	<instagibbs> You'd also have to somehow track that you're no longer asking for any parents of an orphan in order to EraseOrphanOfPeer?
17:43 	<marcofleon> yeah i was thinking it made sense with GetCandidatePeers. Having another tracker to separate out the orphan reso process
17:46 	<glozow> will think about this idea some more
17:47 	<dergoegge> me too 👍
17:47 	<glozow> I think it's possible it works? My main questions would be (1) what is the cost of having a second tracker? Because it's the perfect data structure for this. (2) how do we make it extensible to package relay still.
17:48 	<instagibbs> imo the cost is mostly an additional structure lying around that needs to stay in sync with other things
17:48 	<dergoegge> 1) if we don't need it then it's just extra complexity 2) fair!
17:48 	<marcofleon> The fact that there are candidates that be added or not added to that new tracker is why it made sense to me in the first place I guess is what i was saying
17:48 	<marcofleon> can be*
17:49 	<glozow> (1) shoving it into the existing tracker but needing to have extra logic could also be complexity!
17:49 	<dergoegge> well in my mind it wouldn't need extra logic but I might be wrong, need to think more
17:49 	<instagibbs> proof of code for this I think..
17:49 	<glozow> but yeah, I don't like that we need to ensure m_orphanage and m_orphan_resolution_tracker are in sync. that's super fair
17:49 	<dergoegge> frfr
17:50 	<glozow> yeah let's see what the code would look like
17:50 	<marcofleon> fr
17:50 	<glozow> fr r r
```

</details>

glozow commented:
Thanks @dergoegge, will spend some time thinking about whether there are any problems with it. Perhaps we can do this easier version now and cross the orphan reso tracker bridge when we need to build the 2-step orphan resolution protocol (info and then transactions).

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -174,10 +180,27 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
 
 bool TxDownloadManagerImpl::AddTxAnnouncement(NodeId peer, const GenTxid& gtxid, std::chrono::microseconds now, bool p2p_inv)
```
added a commit to clean this up

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -174,10 +180,27 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
+        return true;
+    }
+
+    const bool already_had{AlreadyHaveTx(gtxid, /*include_reconsiderable=*/true)};
```
cleaned up

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -174,10 +180,27 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
+
     // If this is an inv received from a peer and we already have it, we can drop it.
     // If this is a request for the parent of an orphan, we don't drop transactions that we already have. In particular,
     // we *do* want to request parents that are in m_lazy_recent_rejects_reconsiderable, since they can be CPFP'd.
```
cleaned up

glozow commented on src/txorphanage.cpp:
```diff
@@ -12,12 +12,20 @@
+        const auto ret = it->second.announcers.insert(peer);
+        if (ret.second) {
+            LogDebug(BCLog::TXPACKAGES, "added peer=%d as announcer of orphan tx %s\n", peer, wtxid.ToString());
+        }
```
largely agree, done

glozow committed:
[cleanup] remove p2p_inv from AddTxAnnouncement

This param is no longer needed since orphan parent requests are added to
the TxRequestTracker directly.

glozow commented on src/txorphanage.cpp:
```diff
@@ -16,8 +16,16 @@ bool TxOrphanage::AddTx(const CTransactionRef& tx, NodeId peer, const std::vecto
+        if (ret.second) {
+            LogDebug(BCLog::TXPACKAGES, "added peer=%d as announcer of orphan tx %s\n", peer, wtxid.ToString());
+        }
+        // Even if an announcer was added, no new orphan entry was created.
```
We haven't necessarily added an announcer. But I updated this comment.

glozow commented on src/txorphanage.cpp:
```diff
@@ -124,36 +124,41 @@ void TxOrphanage::EraseForPeer(NodeId peer)
             std::map<Wtxid, OrphanTx>::iterator maybeErase = iter++;
             if (maybeErase->second.nTimeExpire <= nNow) {
-                nErased += EraseTx(maybeErase->second.tx->GetWitnessHash());
+                const auto& wtxid = maybeErase->second.tx->GetWitnessHash();
```
switched

glozow commented on src/txorphanage.cpp:
```diff
@@ -124,36 +124,41 @@ void TxOrphanage::EraseForPeer(NodeId peer)
         // Evict a random orphan:
         size_t randompos = rng.randrange(m_orphan_list.size());
-        EraseTx(m_orphan_list[randompos]->second.tx->GetWitnessHash());
+        const auto& wtxid = m_orphan_list[randompos]->second.tx->GetWitnessHash();
```
done

glozow commented on src/txorphanage.cpp:
```diff
@@ -152,6 +183,12 @@ bool TxOrphanage::HaveTx(const Wtxid& wtxid) const
     return m_orphans.count(wtxid);
 }
 
+bool TxOrphanage::HaveTxAndPeer(const Wtxid& wtxid, NodeId peer) const
```
changed

glozow commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -239,20 +239,15 @@ def test_orphan_consensus_failure(self):
         assert tx_orphan_bad_wit.rehash() not in node_mempool
 
-        # 5. Peer that sent a consensus-invalid transaction should be disconnected.
-        bad_orphan_sender.wait_for_disconnect()
```
Good point, added a line where this sender sends the parent too

glozow commented on src/txorphanage.h:
```diff
@@ -80,7 +90,8 @@ class TxOrphanage {
     struct OrphanTxBase {
         CTransactionRef tx;
-        NodeId fromPeer;
+        /** Peers added with AddTx or AddAnnouncer. */
```
Correct. I think this is not too misleading? Added "unused for now" in the commit messages.

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -624,6 +625,74 @@ def test_max_orphan_amount(self):
+        # Announces after tx is sent
+        peer_late_announcer = node.add_p2p_connection(PeerTxRelayer())
+
+        # Both peers send invs for the orphan, so the node an expect both to know its ancestors.
```
done

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -308,7 +308,7 @@ node::RejectedTxTodo TxDownloadManagerImpl::MempoolRejectedTx(const CTransaction
     bool add_extra_compact_tx{first_time_failure};
     // Hashes to pass to AddKnownTx later
-    std::vector<uint256> unique_parents;
+    std::vector<Txid> unique_parents;
```
done

glozow reviewed (commented):
Changed to the new approach, cleaned up a bunch of things, addressed most comments. I think I have 1 or 2 more to get to, and need to write an RPC test.

instagibbs commented on test/functional/p2p_1p1c_network.py:
```diff
@@ -143,20 +143,14 @@ def run_test(self):
 
         self.log.info("Wait for mempools to sync")
-        self.sync_mempools(timeout=20)
+        self.sync_mempools(timeout=62)
```
62 is very magical looking, can we do better?

glozow replied:
deleted now, I think it was a hack for some intermittent thing that I should have removed a while ago

instagibbs reviewed (commented):
took another pass with the major new commit f96df3d60a943b87eecd3a6915fcace0eee2838f

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -174,9 +175,30 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
+    // - exists in orphanage
+    // - peer can be an orphan resolution candidate
+    if (p2p_inv && gtxid.IsWtxid()) {
+        if (auto parent_txids{m_orphanage.GetParentTxids(Wtxid::FromUint256(gtxid.GetHash()))}) {
```
> also, should we ever have something in our orphanage that doesn't have a parent txid?

Yeah we shouldn't ever have that situation. However I like the `std:optional<X> GetX()` interface convention.

glozow committed:
[functional test] getorphantxs reflects multiple announcers

glozow commented on test/functional/p2p_opportunistic_1p1c.py:
```diff
@@ -279,20 +274,17 @@ def test_parent_consensus_failure(self):
-
         self.log.info("Check that fake parent does not cause orphan to be deleted and real package can still be submitted")
-        # 6. Child-sending should not have been punished and the orphan should remain in orphanage.
+        # 5. Child-sending should not have been punished and the orphan should remain in orphanage.
```
Done, and added a multi-announcer test to rpc_orphans.py

instagibbs commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -97,6 +97,7 @@ void TxDownloadManagerImpl::ActiveTipChange()
 
 void TxDownloadManagerImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock)
 {
+    // Orphanage may include transactions conflicted by this block.
```
micro-nit: this being added in a32d396ef27d626e0c60e7ee8a74cbe3d918acb1 is weird

instagibbs commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -204,6 +226,35 @@ bool TxDownloadManagerImpl::AddTxAnnouncement(NodeId peer, const GenTxid& gtxid,
+        if (m_txrequest.Count(nodeid) >= MAX_PEER_TX_ANNOUNCEMENTS) return std::nullopt;
+    }
+
+    std::chrono::seconds delay{0s};
```
take or leave suggestion: instead of copying and pasting, make the delay computation re-usable?
```
diff --git a/src/node/txdownloadman_impl.cpp b/src/node/txdownloadman_impl.cpp
index 41d1badadf..d6e44c321b 100644
--- a/src/node/txdownloadman_impl.cpp
+++ b/src/node/txdownloadman_impl.cpp
@@ -171,10 +171,28 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
         m_peer_info.erase(it);
     }
 
 }
 
+std::chrono::seconds TxDownloadManagerImpl::GetTxRequestDelay(NodeId peer, const TxDownloadConnectionInfo& info, const GenTxid& gtxid)
+{
+    // Decide the TxRequestTracker parameters for this announcement:
+    // - "preferred": if fPreferredDownload is set (= outbound, or NetPermissionFlags::NoBan permission)
+    // - "reqtime": current time plus delays for:
+    //   - NONPREF_PEER_TX_DELAY for announcements from non-preferred connections
+    //   - TXID_RELAY_DELAY for txid announcements while wtxid peers are available
+    //   - OVERLOADED_PEER_TX_DELAY for announcements from peers which have at least
+    //     MAX_PEER_TX_REQUEST_IN_FLIGHT requests in flight (and don't have NetPermissionFlags::Relay).
+    auto delay{0s};
+    if (!info.m_preferred) delay += NONPREF_PEER_TX_DELAY;
+    if (!gtxid.IsWtxid() && m_num_wtxid_peers > 0) delay += TXID_RELAY_DELAY;
+    const bool overloaded = !info.m_relay_permissions && m_txrequest.CountInFlight(peer) >= MAX_PEER_TX_REQUEST_IN_FLIGHT;
+    if (overloaded) delay += OVERLOADED_PEER_TX_DELAY;
+
+    return delay;
+}
+
 bool TxDownloadManagerImpl::AddTxAnnouncement(NodeId peer, const GenTxid& gtxid, std::chrono::microseconds now, bool p2p_inv)
 {
     // If this is an orphan we are trying to resolve, consider this peer as a orphan resolution candidate instead.
     // - received as an p2p inv
     // - is wtxid matching something in orphanage
@@ -206,24 +224,12 @@ bool TxDownloadManagerImpl::AddTxAnnouncement(NodeId peer, const GenTxid& gtxid,
     const auto& info = it->second.m_connection_info;
     if (!info.m_relay_permissions && m_txrequest.Count(peer) >= MAX_PEER_TX_ANNOUNCEMENTS) {
         // Too many queued announcements for this peer
         return false;
     }
-    // Decide the TxRequestTracker parameters for this announcement:
-    // - "preferred": if fPreferredDownload is set (= outbound, or NetPermissionFlags::NoBan permission)
-    // - "reqtime": current time plus delays for:
-    //   - NONPREF_PEER_TX_DELAY for announcements from non-preferred connections
-    //   - TXID_RELAY_DELAY for txid announcements while wtxid peers are available
-    //   - OVERLOADED_PEER_TX_DELAY for announcements from peers which have at least
-    //     MAX_PEER_TX_REQUEST_IN_FLIGHT requests in flight (and don't have NetPermissionFlags::Relay).
-    auto delay{0us};
-    if (!info.m_preferred) delay += NONPREF_PEER_TX_DELAY;
-    if (!gtxid.IsWtxid() && m_num_wtxid_peers > 0) delay += TXID_RELAY_DELAY;
-    const bool overloaded = !info.m_relay_permissions && m_txrequest.CountInFlight(peer) >= MAX_PEER_TX_REQUEST_IN_FLIGHT;
-    if (overloaded) delay += OVERLOADED_PEER_TX_DELAY;
 
-    m_txrequest.ReceivedInv(peer, gtxid, info.m_preferred, now + delay);
+    m_txrequest.ReceivedInv(peer, gtxid, info.m_preferred, now + GetTxRequestDelay(peer, info, gtxid));
 
     return false;
 }
 
 std::optional<std::chrono::seconds> TxDownloadManagerImpl::OrphanResolutionCandidate(NodeId nodeid, const Wtxid& orphan_wtxid)
@@ -240,21 +246,12 @@ std::optional<std::chrono::seconds> TxDownloadManagerImpl::OrphanResolutionCandi
         // orphan resolution involves at least 1 transaction request which may or may not be
         // currently tracked in m_txrequest, so we include that in the count.
         if (m_txrequest.Count(nodeid) >= MAX_PEER_TX_ANNOUNCEMENTS) return std::nullopt;
     }
 
-    std::chrono::seconds delay{0s};
-    if (!info.m_preferred) delay += NONPREF_PEER_TX_DELAY;
-    // The orphan wtxid is used, but resolution entails requesting the parents by txid. Sometimes
-    // parent and child are announced and thus requested around the same time, and we happen to
-    // receive child sooner. Waiting a few seconds may allow us to cancel the orphan resolution
-    // request if the parent arrives in that time.
-    if (m_num_wtxid_peers > 0) delay += TXID_RELAY_DELAY;
-    const bool overloaded = !info.m_relay_permissions && m_txrequest.CountInFlight(nodeid) >= MAX_PEER_TX_REQUEST_IN_FLIGHT;
-    if (overloaded) delay += OVERLOADED_PEER_TX_DELAY;
-
-    return delay;
+    // We're fetching parents via txid; so pass in as a txid
+    return GetTxRequestDelay(nodeid, info, GenTxid::Txid(orphan_wtxid));
 }
 
 std::vector<GenTxid> TxDownloadManagerImpl::GetRequestsToSend(NodeId nodeid, std::chrono::microseconds current_time)
 {
     std::vector<GenTxid> requests;
diff --git a/src/node/txdownloadman_impl.h b/src/node/txdownloadman_impl.h
index 560802ab5d..0bd673e956 100644
--- a/src/node/txdownloadman_impl.h
+++ b/src/node/txdownloadman_impl.h
@@ -158,10 +158,12 @@ public:
     bool AlreadyHaveTx(const GenTxid& gtxid, bool include_reconsiderable);
 
     void ConnectedPeer(NodeId nodeid, const TxDownloadConnectionInfo& info);
     void DisconnectedPeer(NodeId nodeid);
 
+    std::chrono::seconds GetTxRequestDelay(NodeId peer, const TxDownloadConnectionInfo& info, const GenTxid& gtxid);
+
     /** Consider adding this tx hash to txrequest. Should be called whenever a new inv has been received.
      * Also called internally when a transaction is missing parents so that we can request them.
      */
     bool AddTxAnnouncement(NodeId peer, const GenTxid& gtxid, std::chrono::microseconds now, bool p2p_inv);
 ```

instagibbs commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -354,26 +405,45 @@ node::RejectedTxTodo TxDownloadManagerImpl::MempoolRejectedTx(const CTransaction
+                // that if a peer is overloading us with invs and orphans, they will eventually not be
+                // able to add any more transactions to the orphanage.
+                add_orphan_reso_candidate(ptx, unique_parents, nodeid, now);
+                for (const auto& candidate : m_txrequest.GetCandidatePeers(wtxid)) {
```
question: since `GetCandidatePeers` is only used in this context, can we just have it take a transaction ref and fetch both txid and wtxid under the hood?

instagibbs reviewed (approved):
ACK 40abc3f4f1a2b7c6da657f2cdffd2b61f5b1b91a

non-blocking suggestions only

mzumsande reviewed (commented):
Concept ACK
Did a first round of review, mostly comment nits (it's been a while since I last looked at the orphanage).

instagibbs commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -204,6 +226,35 @@ bool TxDownloadManagerImpl::AddTxAnnouncement(NodeId peer, const GenTxid& gtxid,
+        // existing behavior: drop if we are tracking too many invs for this peer already. Each
+        // orphan resolution involves at least 1 transaction request which may or may not be
+        // currently tracked in m_txrequest, so we include that in the count.
+        if (m_txrequest.Count(nodeid) >= MAX_PEER_TX_ANNOUNCEMENTS) return std::nullopt;
```
The fuzzer should theoretically hit this if so, though the limit is probably high enough to make this hard to hit.

glozow commented on src/txorphanage.h:
```diff
@@ -84,6 +89,8 @@ class TxOrphanage {
     struct OrphanTx : public OrphanTxBase {
         size_t list_pos;
+        /** Txids of the missing parents to request. Determined by peerman. */
+        std::vector<Txid> parent_txids;
```
Oh, I just had a thought that maybe it's kind of bad that we're locked in to the `parent_txids` we started with... even if we're able to remove some, what happens if a new parent goes missing?

1. Get tx that spends [A, B, C]. We are missing inputs A and B. Store in tx orphanage. Request A and B from peers...
2. C gets replaced* or falls out of mempool somehow
3. We keep only requesting A and B. We receive and accept them.
4. AddChildrenToWorkSet... C is still an orphan. We can't figure out how to resolve it properly though.

So maybe we do this, but also need to update `parent_txids` on every `MempoolRejectedTx` `TX_MISSING_INPUTS` as well.

instagibbs replied:
Note that also `A` could immediately be replaced after fetching it, before `B` is received. There are probably a bunch of ways this and other ancestor fetching can fail.

Even if we manage to recover the entire ancestry tree (like in this example ABC+tx), if C was replaced, we'd likely be relying on general package RBF to have the tx successfully submitted.

> That's a good idea! Especially for 1p1cs 🤔 . We could update parent_txids at the same time as AddChildrenToWorkSet and only add them to workset when empty.

This seems reasonable on its own.

glozow replied:
Perhaps we start with recalculating every time and consider doing this later? I agree it doesn't seem too expensive.

instagibbs commented on src/txorphanage.cpp:
```diff
@@ -135,13 +157,17 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx)
-                LogDebug(BCLog::TXPACKAGES, "added %s (wtxid=%s) to peer %d workset\n",
-                         tx.GetHash().ToString(), tx.GetWitnessHash().ToString(), elem->second.fromPeer);
+                // Belt and suspenders, each orphan should always have at least 1 announcer.
+                if (!Assume(!elem->second.announcers.empty())) break;
```
Why break? Do we not want to try another orphan? We keep trying the other tx's outpoints regardless.

JeremiahR commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -172,12 +173,33 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
+                LogDebug(BCLog::TXPACKAGES, "added peer=%d as a candidate for resolving orphan %s\n", peer, gtxid.GetHash().ToString());
+            }
+
+            // Return even if the peer isn't an orphan resolution candidate. This would be caught by AlreadhHaveTx.
```
typo 
```suggestion
            // Return even if the peer isn't an orphan resolution candidate. This would be caught by AlreadyHaveTx.
```

glozow committed:
[txorphanage] add GetTx so that orphan vin can be read

glozow committed:
[refactor] move creation of unique_parents to helper function

This function will be reused in a later commit.

glozow committed:
[unit test] multiple orphan announcers

glozow commented:
Thanks for the reviews @mzumsande @instagibbs! Just got back from holiday, addressed all the comments. Biggest change was recalculating missing parents each time we add a new orphan reso candidate. I also wrote a test for https://github.com/bitcoin/bitcoin/pull/31397#discussion_r1881018172

sipa reviewed (commented):
I wonder if it's possible to avoid keeping track of the announcers in `m_orphans`, as I think it should match `m_txrequest.GetCandidatePeers(orphan_tx)`?

glozow commented:
> I wonder if it's possible to avoid keeping track of the announcers in `m_orphans`, as I think it should match `m_txrequest.GetCandidatePeers(orphan_tx)`?

That's true immediately after receiving the orphan, but we delete the entries from `m_txrequest` once we've added this tx to the orphanage - `m_txrequest` only stores what we haven't successfully downloaded yet. So I think we need to store the announcers somewhere else to remember that information, remove candidates that fail to resolve the orphan, and add new candidates if they announce the tx. `TxOrphanage` seems like the best place for it.

sipa reviewed (commented):
Concept ACK, some nits/questions:

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -174,9 +174,35 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
+            });
+
+            if (auto delay{OrphanResolutionCandidate(peer, Wtxid::FromUint256(gtxid.GetHash()), unique_parents.size())}) {
+                m_orphanage.AddAnnouncer(Wtxid::FromUint256(gtxid.GetHash()), peer);
```
Hm right, I don't think we can assume that it's never empty - I've added an exit case

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -363,26 +418,42 @@ node::RejectedTxTodo TxDownloadManagerImpl::MempoolRejectedTx(const CTransaction
+
+                auto add_orphan_reso_candidate = [&](const CTransactionRef& orphan_tx, std::vector<Txid> unique_parents, NodeId nodeid, std::chrono::microseconds now) {
+                    const auto& wtxid = orphan_tx->GetWitnessHash();
+                    if (auto delay{OrphanResolutionCandidate(nodeid, wtxid, unique_parents.size())}) {
```
Yes, it is similar. I'm open to code diff suggestions.

I tried a few times to refactor but found some things a bit ugly to work around: we force the `IsWtxid` check to be false since we're requesting parents by txid, and the number of requests is not always 1 (also see https://github.com/bitcoin/bitcoin/pull/31397#discussion_r1901875482). In the future, the function may deviate when we add more per-peer orphan resolution limits and/or another way to do orphan resolution.

glozow replied:
Ah, makes sense - I thought you meant a different bit of code. Added in #31666

instagibbs commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -174,9 +174,37 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
+    // - exists in orphanage
+    // - peer can be an orphan resolution candidate
+    if (p2p_inv && gtxid.IsWtxid()) {
+        if (auto orphan_tx{m_orphanage.GetTx(Wtxid::FromUint256(gtxid.GetHash()))}) {
```
```Suggestion
        const auto wtxid{Wtxid::FromUint256(gtxid.GetHash()))};
        if (auto orphan_tx{m_orphanage.GetTx(wtxid)}) {
```
and use `wtxid` below in the other two locations

glozow replied:
Added in #31666

marcofleon reviewed (commented):
Code review ACK 86d7135e36efd39781cf4c969011df99f0cbb69d

Ran all the relevant tests, including the fuzz tests for about 50 cpu hours each. The implementation here of keeping track of multiple peers that announce an orphan looks good to me. It makes sense to not be locked in to only one peer for orphan resolution, as that peer could disconnect (in which case we lose the orphan) or be malicious.

Just left some non-blocking nits below.

Also quick question for my own understanding. Is it true that in the honest case, orphan resolution will probably happen with the first peer that announced it to us? The peer should have the ancestors, so those txs should be sent shortly after the orphan tx was received by us, right?

instagibbs reviewed (approved):
reACK 86d7135e36efd39781cf4c969011df99f0cbb69d

feel completely ok to ignore all my comments

Major change is recomputing the missing parents per announcement of orphan, rather than a static list for the lifetime of the orphan when first received, and also correctly accounting for those missing parents when deciding when to rate-limit the announcements for a specific peer..

instagibbs commented:
@marcofleon 

>Also quick question for my own understanding. Is it true that in the honest case, orphan resolution will probably happen with the first peer that announced it to us?

It Depends(TM). If the first announcer was an inbound connection and second was an outbound connection less than 2(?) seconds later, the outbound will be chosen first. see  the functional test case `test_orphan_handling_prefer_outbound`

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -301,14 +329,29 @@ void TxDownloadManagerImpl::MempoolAcceptedTx(const CTransactionRef& tx)
     m_orphanage.EraseTx(tx->GetWitnessHash());
 }
 
+std::vector<Txid> TxDownloadManagerImpl::GetUniqueParents(const CTransaction& tx)
```
Ah I see what you mean, yes!

dergoegge reviewed (approved):
Code review ACK 86d7135e36efd39781cf4c969011df99f0cbb69d

Other reviewers already mentioned the nits that I would have had as well. I think they are fine to address in a follow up.

mzumsande reviewed (commented):
ACK 86d7135e36efd39781cf4c969011df99f0cbb69d

I reviewed the code and tested it a bit on mainnet with some extra logging: The large majority of orphans gets resolved, a few orphans get stuck for 20 minutes if all of our candidates send `NOTFOUND`s for the parent requests, presumably because the parent got removed from their mempools  (this has been mentioned in https://github.com/bitcoin/bitcoin/pull/31397#issuecomment-2523122642 ).

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -172,12 +172,39 @@ void TxDownloadManagerImpl::DisconnectedPeer(NodeId nodeid)
+                return AlreadyHaveTx(GenTxid::Txid(txid), /*include_reconsiderable=*/false);
+            });
+
+            if (unique_parents.empty()) return true;
```
Done in followup, and also think we can `EraseTx` here. See #31666

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -58,6 +58,10 @@ def wrapper(self):
             self.nodes[0].disconnect_p2ps()
             self.nodes[0].bumpmocktime(LONG_TIME_SKIP)
+            # Check that mempool and orphanage have been cleared
+            assert_equal(0, len(self.nodes[0].getorphantxs()))
```
yes, I think it needs to be a `wait_until`

glozow replied:
#31675

glozow commented on src/txorphanage.cpp:
```diff
@@ -135,13 +158,17 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx)
+                for (const auto announcer: elem->second.announcers) {
+                    // Get this source peer's work set, emplacing an empty set if it didn't exist
+                    // (note: if this peer wasn't still connected, we would have removed the orphan tx already)
+                    std::set<Wtxid>& orphan_work_set = m_peer_work_set.try_emplace(announcer).first->second;
```
Good point. I seem to recall drafting a version that randomly selected 1 peer. That maybe saves us from a peer purposefully disconnecting (since they don't know whether we assigned them), but doesn't have this redundancy problem? What do you think?

mzumsande replied:
Yes, taking a random peer would be ok.
Though I kinda like the idea of calling `ProcessOrphanTx()` in `FinalizeNode()` in general, even if just 1 peer is involved - this is scheduled work that is assigned to but doesn't need any input from the peer, so there is no reason not do it just because the peer decides to disconnect at the wrong time.

glozow replied:
I like that idea too. But what if there is a large amount of orphans left? There shouldn't be a ton, but I don't know how to feel about a peer being granted more than the normal amount of compute because they are going to be disconnected.

mzumsande replied:
Right, with`ProcessOrphanTx()` doing at most one actual validation, it would probably not be good to call it in a loop until all outstanding work is done. Maybe calling it just once (as if it was this peer's next turn) would already help in most cases, but to find out it'd probably be best to get some empirical data from mainnet.

glozow commented on src/txrequest.cpp:
```diff
@@ -574,6 +574,17 @@ class TxRequestTracker::Impl {
         }
     }
 
+    std::vector<NodeId> GetCandidatePeers(const uint256& txhash) const
```
Added in #31666

glozow commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -320,13 +366,7 @@ node::RejectedTxTodo TxDownloadManagerImpl::MempoolRejectedTx(const CTransaction
-            }
-            std::sort(unique_parents.begin(), unique_parents.end());
-            unique_parents.erase(std::unique(unique_parents.begin(), unique_parents.end()), unique_parents.end());
+            unique_parents = GetUniqueParents(tx);
```
I feel like they're about the same, so going to leave as is

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -626,15 +630,181 @@ def test_max_orphan_amount(self):
+        node.sendrawtransaction(tx_replacer_BC["hex"])
+        assert tx_replacer_BC["txid"] in node.getrawmempool()
+        node.sendrawtransaction(tx_replacer_C["hex"])
+        assert tx_replacer_BC["txid"] not in node.getrawmempool()
```
Added in #31666

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -626,15 +630,181 @@ def test_max_orphan_amount(self):
+        assert tx_replacer_BC["txid"] not in node.getrawmempool()
+        assert tx_replacer_C["txid"] in node.getrawmempool()
+
+        # Second peer is an additional announcer for this orphan
```
Added in #31666

glozow commented on test/functional/p2p_orphan_handling.py:
```diff
@@ -626,15 +630,181 @@ def test_max_orphan_amount(self):
+
+        # Disconnect peer1. peer2 should become the new candidate for orphan resolution.
+        peer1.peer_disconnect()
+        node.bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
```
Added in #31666

glozow commented:
Followup #31666 wraps up all the comments here except for 2 things which might be a little more involved:

https://github.com/bitcoin/bitcoin/pull/31397#discussion_r1902153685 (because we need to add a txid index)
https://github.com/bitcoin/bitcoin/pull/31397#discussion_r1919108948

glozow commented on src/node/txdownloadman_impl.h:
```diff
@@ -189,6 +189,16 @@ class TxDownloadManagerImpl {
+
+protected:
+    /** Helper for getting deduplicated vector of Txids in vin. */
+    std::vector<Txid> GetUniqueParents(const CTransaction& tx);
```
Right. Not changing for now because it may be useful to recurse in-orphanage ancestors in this function.
